### PR TITLE
New content: Add definition for shape broadcasting

### PR DESCRIPTION
**Pull request by inexorabletash**
This change introduces a new section for Algorithms, following APIs, to collect algorithms referenced throughout the specification.

A section for Broadcasting is introduced, which defines broadcasting shapes and gives an explicit algorithm matching WebNN implementations of NumPy's General Broadcasting Rules. Definitions for "broadcastable" and "unidirectionally broadcastable" are introduced. The previous definition of "broadcast-shapes" is removed in favor of these new algorithms.

For #324, #462, and potentially #523.

***
<a href="https://pr-preview.s3.amazonaws.com/inexorabletash/webnn/pull/534.html" title="Last updated on Feb 13, 2024, 4:40 AM UTC (145bff1)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/534/55183f7...inexorabletash:145bff1.html" title="Last updated on Feb 13, 2024, 4:40 AM UTC (145bff1)">Diff</a>

### CODE REVIEW

**Comment by inexorabletash** [comment]
I'm still pretty hazy on definitions, so please look at this very skeptically. 

Re: introducing a new "Algorithms" section - this is a pattern I've adopted from other specs. I thought I'd give a whirl here and see what everyone thinks. It's easy enough to drop and put these definitions anywhere else to keep the overall spec structure unchanged.

**Comment by inexorabletash** [comment]
Hold off on reviews, I was confusing unidirectional (mentioned in several places in the spec) and bidirectional (mentioned in the Chromium implementation). Fridays, am I right?

**Comment by inexorabletash** [comment]
Okay, this is updated to match the Chromium impl, I think. But... that may not be what's desired @a-sully mentions that unidirectional broadcasting seems specific to ONNX and other frameworks do things to work around this during export.

**Comment by a-sully** [comment]
AFAICT ONNX is an outlier in making a distinction between "unidirectional" and "multidirectional" broadcasting. This distinction is not made by:
- Numpy: https://numpy.org/doc/stable/user/basics.broadcasting.html
- PyTorch: https://pytorch.org/docs/stable/notes/broadcasting.html (which references Numpy)
- Tensorflow/XLA/StableHLO: https://www.tensorflow.org/xla/broadcasting (which also references Numpy, though it has some other restrictions)

The ["unidirectional broadcastable"](https://github.com/onnx/onnx/blob/main/docs/Broadcasting.md#unidirectional-broadcasting) constraint of some ONNX ops (e.g. [`prelu()`](https://github.com/onnx/onnx/blob/main/docs/Operators.md#PRelu)) requires workarounds when exporting from other formats to ONNX - like in this example of using TVM to export PyTorch to ONNX: https://github.com/pytorch/pytorch/issues/70570#issuecomment-1034379620. The folks on that issue discovered that the "unidirectional broadcastable" constraint for `prelu()` was not documented (which has since been fixed) and that there were inconsistencies between ONNX implementations (which... hopefully has also been fixed ¯\\_(ツ)_/¯ )

Numpy's broadcasting rules are the clear standard across the industry and IMHO is the standard that we should expose to the web. It then becomes up to the user agent to ensure that the constraints of the underlying framework (e.g. unidirectional broadcasting for ONNX, [lack of inferred broadcasting specifications for XLA](https://www.tensorflow.org/xla/broadcasting#principles)) are satisfied.

**Comment by wacky6** [comment]
We follow Numpy broadcast rules for ops (in the chromium implementation), I think it's the golden "standard". +1 to adopt this throughout the spec (I didn't realize the spec matmul referred to an implementation defined broadcast 😱).


> @a-sully mentions that unidirectional broadcasting seems specific to ONNX and other frameworks do things to work around this during export

Is ONNX unidirectional broadcast issue easy to workaround? The linked prelu() patch seems "easy enough". 

Should we (from the spec's perspective) expect implementations (e.g. chromium DML backend)to do asid easy workarounds?

File: index.bs
Comment (huningxin) [inline index.bs:2867]: Because it checks size of new shape is equal to input rank, I suppose it can unidirectionally broadcast input shape to new shape.

**Review by huningxin — APPROVED** [review]
LGTM with a comment, thanks much!

**Comment by huningxin** [comment]
@fdwr and @wchao1115 , please also take a look. Thanks!

File: index.bs
Comment (inexorabletash) [reply index.bs]: I pushed a second commit that makes this update. Please take a look?

File: index.bs
Comment (fdwr) [inline index.bs:5990]:
```suggestion
    1. If |condition| is not [=bidirectionally broadcastable=] to |descriptor|.{{MLOperandDescriptor/dimensions}}, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
```

Oops, existing spec typo, as the 3 input tensors (`condition`, `trueValues`, `falseValues`) are actually *multidirectionally* broadcastable with each other, which is functionally equivalent to `bidirectionalBroadcast(a, bidirectionalBroadcast(b, c))`. That's what the reference and implementation do:

https://github.com/webmachinelearning/webnn-baseline/pull/56/files#diff-1ab0e4e354fcd892d24b937bb8aad28636b554a25f248d3433573b57603ea411

```
  const valueShape = getBroadcastShape(trueValuesReshape.shape, falseValuesReshape.shape);
  const outputShape = getBroadcastShape(condition.shape, valueShape);
```

https://chromium-review.googlesource.com/c/chromium/src/+/5039655

```
  const auto value_shape =
      BroadcastShapes(true_value.dimensions, false_value.dimensions, true);
  if (!value_shape) {
    return base::unexpected(
        "The shapes of true_value and false_value are not broadcastable.");
  }

  const auto output_shape =
      BroadcastShapes(condition.dimensions, value_shape.value(), true);
  if (!output_shape) {
    return base::unexpected(
        "The condition shape is not broadcastable to the shape broadcasted "
        "from true_value and false_value.");
  }
```

https://chromium-review.googlesource.com/c/chromium/src/+/5069382

```
// * The shapes of all input operands must be bidirectionally broadcastable. (with each other)
```

File: index.bs
Comment (fdwr) [inline index.bs:2859]: Yeah, it appears the rule about sequence length matching input rank is not necessary, nor is it so in either Chromium implementation or the reference baseline.

File: index.bs
Comment (fdwr) [inline index.bs:6215]:
```suggestion
1. Let |outputShape| be a new [=/list=].
```
Just for added clarity, since `output` often is used for the *tensor*, and consistency with `outputShape` elsewhere in spec.

File: index.bs
Comment (fdwr) [inline index.bs:6237]: Another option that might be clearer and functionally equivalent, because then the loop becomes a more straight-forward foward pass using Append and without the negation, is to just pad A and B upfront (we do this often inside our internal code to avoid going crazy). e.g.

```
1. Let |outputSize| be the maximum of |sizeA| and |sizeB|.
1. Let |output| be a new [=/list=].
1. Let |paddedA| be a copy of |A| with enough leading 1's to pad up to length |outputSize|.
1. Let |paddedB| be a copy of |B| with enough leading 1's to pad up to length |outputSize|.
1. [=list/For each=] |index| in [=the range=] 0 to |outputSize|, exclusive:
    1. Let |dimA| be |paddedA|[|index|] if |index| < |sizeA|, or 1 otherwise.
    1. Let |dimB| be |paddedB|[|index|] if |index| < |sizeB|, or 1 otherwise.
    1. If |dimA| is not equal to |dimB|, and |dimA| is not equal to 1, and |dimB| is not equal to 1, then return failure.
    1. [=list/Append=] the maximum of |dimA| and |dimB| to |output|.
1. Return |output|.
```
🤷‍♂️ (no strong feeling either way)

File: index.bs
Comment (fdwr) [inline index.bs:6214]: It's an error for tensor B.rank > A.rank using undirectional broadcasting (B.rank <= A.rank is fine). Propose:
```suggestion
1. Let |sizeB| be the [=list/size=] of |B|.
1. If |sizeB| > |sizeA|, then return failure.
```

**Review by fdwr — COMMENTED** [review]
Thank you Josh.

**Comment by fdwr** [comment]
> The unidirectional broadcastable constraint of some ONNX ops (e.g. prelu() requires workarounds when exporting from other formats ...

Huh, I would have presumed `prelu` was bidirectionally broadcastable too. TIL... For operator behavior changes, I'd like it to be separate from this CR that's just adding definitions.

> Is ONNX unidirectional broadcast issue easy to workaround? The linked prelu() patch seems "easy enough".

Yep, easy with a little reshape of leading 1's.

> Should we (from the spec's perspective) expect implementations (e.g. chromium DML backend)to do asid easy workarounds?

Whichever broadcasting behavior we have for prelu, unidirectional or bidirectional, it's trivial either way for backends to insert leading 1's as needed (which the Chromium DML backend already does for DML, since it expects elementwise tensor descriptions of each input to have been massaged to consistent rank first). In the worst case, backends can utilize whatever logic they have for `expand` to pre-expand tensors before forwarding them along.

File: index.bs
Comment (inexorabletash) [reply index.bs:6238]: Yeah I get too clever sometimes. Suggestion SGTM

**Comment by inexorabletash** [comment]
History was getting messy - I force-pushed a new squashed commit. It rolls in #564 to get the build working, but I'll rebase once that merges.

**Comment by fdwr** [comment]
>  I force-pushed a new squashed commit. 

**Josh**: 🤔 Shoot, I'm unsure if I merged your layernorm CR via squash commit like I was supposed. Except larger branch integrations where we want to retain history, that's the general policy for cleaner history. Then the CR author doesn't need to ever explicitly squash their changes, because force pushes in GitHub yield this unfortunate result for reviewers to minimize the review delta since last time...

image

**Review by fdwr — APPROVED** [review]
LGTM. 👞✨

**Comment by inexorabletash** [comment]
No worries... I tend to squash locally, force-push, and always review the full diff anyway to see what the actual change is going to be. I'll try to refrain from squashing/force-pushing here though, going forward!

File: index.bs
Comment (fdwr) [reply index.bs]: Appears to me this comment is resolved, and Ningxin already signed off for the rest. Given he is IIRC out \~Feb 10th to 24th for the lunar new year, I propose completing this, and having Ningxin ask for any smaller changes in a later delta.

File: index.bs
Comment (fdwr) [inline index.bs:4721]: If A is the first parameter and B is the second parameter (where A is broadcasted to B), then based on the definition below, these two should be swapped so `slope` is broadcasted to `input`, consistent with `expand`:

```suggestion
    1. Set |descriptor|.{{MLOperandDescriptor/dimensions}} to the result of [=unidirectionally broadcasting the shapes=] |slope|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}} and |input|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}}.
```

**Review by fdwr — APPROVED** [review]
One more tiny fix after re-review, which I'll just (per your earlier permission) submit and complete this. Thanks Josh.

**Pull request merged**
